src/dotty/tools/dotc/reporting/ConsoleReporter.scala

```diff
@@ -4,53 +4,123 @@ package reporting
+    if (pos.exists) Blue({
+      val file = pos.source.file.toString
+      val errId =
+        if (message.errorId != NoExplanation.ID)
+          s"[E${"0" * (3 - message.errorId.toString.length) + message.errorId}] "
+        else ""
+      val kind =
+        if (message.kind == "") diagnosticLevel
+        else s"${message.kind} $diagnosticLevel"
+      val prefix = s"-- ${errId}${kind}: $file "
+
+      prefix +
+      ("-" * math.max(ctx.settings.pageWidth.value - stripColor(prefix).length, 0))
+    }).show else ""
+
   /** Prints the message with the given position indication. */
-  def printMessageAndPos(msg: String, pos: SourcePosition)(implicit ctx: Context): Unit = {
-    val posStr = if (pos.exists) s"$pos: " else ""
-    printMessage(posStr + msg)
+  def printMessageAndPos(msg: Message, pos: SourcePosition, diagnosticLevel: String)(implicit ctx: Context): Unit = {
+    printMessage(posStr(pos, diagnosticLevel, msg))
     if (pos.exists) {
-      printSourceLine(pos)
-      printColumnMarker(pos)
-    }
+      val (src, offset) = sourceLine(pos)
+      val marker = columnMarker(pos, offset)
+      val err = errorMsg(pos, msg.msg, offset)
+
+      printMessage(List(src, marker, err).mkString("\n"))
+    } else printMessage(msg.msg)
   }
 
-  override def doReport(d: Diagnostic)(implicit ctx: Context): Unit = d match {
-    case d: Error =>
-      printMessageAndPos(s"error: ${d.message}", d.pos)
-      if (ctx.settings.prompt.value) displayPrompt()
-    case d: ConditionalWarning if !d.enablingOption.value =>
-    case d: MigrationWarning =>
-      printMessageAndPos(s"migration warning: ${d.message}", d.pos)
-    case d: Warning =>
-      printMessageAndPos(s"warning: ${d.message}", d.pos)
-    case _ =>
-      printMessageAndPos(d.message, d.pos)
+  def printExplanation(m: Message)(implicit ctx: Context): Unit = {
+    printMessage(hl"""|
+                      |${Blue("Explanation")}
+                      |${Blue("===========")}""".stripMargin)
+    printMessage(m.explanation)
+  }
+
+  override def doReport(m: MessageContainer)(implicit ctx: Context): Unit = {
+    m match {
+      case m: Error =>
+        printMessageAndPos(m.contained, m.pos, "Error")
+        if (ctx.settings.prompt.value) displayPrompt()
+      case m: ConditionalWarning if !m.enablingOption.value =>
+      case m: FeatureWarning =>
+        printMessageAndPos(m.contained, m.pos, "Feature Warning")
+      case m: DeprecationWarning =>
+        printMessageAndPos(m.contained, m.pos, "Deprecation Warning")
+      case m: UncheckedWarning =>
+        printMessageAndPos(m.contained, m.pos, "Unchecked Warning")
+      case m: MigrationWarning =>
+        printMessageAndPos(m.contained, m.pos, "Migration Warning")
+      case m: Warning =>
+        printMessageAndPos(m.contained, m.pos, "Warning")
+      case m: Info =>
+        printMessageAndPos(m.contained, m.pos, "Info")
+    }
+
+    if (ctx.shouldExplain(m)) printExplanation(m.contained)
```

Comment:
Good idea! I can intelligently look if there is an explanation at all and suggest it if it exists.